### PR TITLE
Add `slice_rows` to interchange protocol

### LINKED ISSUE #204: How to slice rows? Can it fit into the interchange, or is the standard required?
Came across this yesterday: https://github.com/microsoft/vscode-jupyter/pull/13951
```python
        elif _VSCODE_builtins.hasattr(df, "to_pandas"):
            df = df.to_pandas().iloc[start:end]
```

This could be improved if they only needed to convert to pandas the part of the data which they needed, i.e.:
```python
elif hasattr(df, "__dataframe__"):
    df = pd.api.interchange.from_dataframe(df.__dataframe__().slice_rows(start, end))
```

Or, if the Standard were actually available:
```python
elif hasattr(df, "__dataframe__") and hasattr(df, "__dataframe_standard__"):
    df = pd.api.interchange.from_dataframe(df.__dataframe_standard().slice_rows(start, end).dataframe)
```

So, could this fit into the interchange protocol, or would filtering by rows only be possible for libraries implementing the standard?

It just seems a bit much to require the standard (especially if a separate compatibility package is required for it) if all people need to do is slice rows

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #204

### CODE REVIEW

**Comment by kkraus14** [comment]
I think this current implementation leaves a lot to be desired. I.E. in the case of something like pandas or other dataframe library that doesn't use the Arrow memory layout under the hood, they'd presumably materialize arrow on the `__dataframe__` call and then have to slice the arrow format memory, which if containing strings or has a step size, isn't free. This is already potentially a problem in selecting columns as well, so I guess this inefficiency is nothing new?

Additionally, it makes it a bit hard to reason about when the producer vs when the consumer should do row selection. I.E. if Polars is consuming data from say PyArrow, I imagine Polars would rather handle row slicing itself (assuming you'll hit a situation where it's not pure pointer arithmetic). Now in the situation of Pandas consuming data from say Polars, you'd probably want Polars to handle the row slicing.

Arrow interchange protocols handle the slicing case (ignoring step size) by allowing specifying an offset and a size. Maybe we can do something similar here?

**Comment by MarcoGorelli** [comment]
sounds good, thanks

File: protocol/dataframe_protocol.py [hunk 1]
```diff
@@ -476,6 +478,28 @@ def select_columns_by_name(self, names: Sequence[str]) -> "DataFrame":
         """
         pass
 
+    @abstractmethod
+    def slice_rows(
+        self,
+        offset: int,
+        length: int | None,
```
Comment (kkraus14) [inline protocol/dataframe_protocol.py:485]: Should we call this `size` to make it consistent with the property we have on the `Column` objects? https://github.com/data-apis/dataframe-api/blob/main/protocol/dataframe_protocol.py#L203

**Comment by kkraus14** [comment]
Do we expect / want to encourage developers using dataframe libraries to explicitly call `__dataframe__` themselves as opposed to using `libraryx.from_dataframe(...)`? It feels a bit funky to me currently that we go from say:
```
pl_df = ...  # My polars dataframe
pdf = pandas.from_dataFrame(pl_df)
```

to:
```
pl_df = ...  # My polars dataframe
pdf = pandas.from_dataframe(pl_df.__dataframe__().select_columns(...).slice_rows(...))
```

My 2c is that this is just highlighting the lack of standard API here and that the experience should be something along the lines of (ignoring API names for column selection and row slicing):
```
pl_df = ...  # My polars dataframe
pdf = pandas.from_dataframe(pl_df.cols(...).slice_rows(...))
```

**Comment by kkraus14** [comment]
Would be good to have others chime in here given this interchange protocol is already being adopted where we probably don't want to introduce something and later decide to change / remove it.

**Comment by MarcoGorelli** [comment]
It's what plotly already does to not have to convert the entire dataframe

**Comment by MarcoGorelli** [comment]
Any updates here please?

This is the only thing I plan to try adding to the interchange protocol, promised

I think of the interchange protocol as being useful to converting between libraries and doing some preselection in a standardised way:
- select columns (currently possible)
- select rows (not possible)

cc @rgommers @jorisvandenbossche

**Comment by MarcoGorelli** [comment]
gentle ping

(would really like to get this in for pandas 3.0 tbh)

**Comment by anmyachev** [comment]
> I think of the interchange protocol as being useful to converting between libraries and doing some preselection in a standardised way:

> select columns (currently possible)
> select rows (not possible)

The ability to select subset rows in addition to selecting columns seems harmonious.

Implementation in Modin should not be a problem.

+1

**Comment by MarcoGorelli** [comment]
Any updates please?

**Comment by MarcoGorelli** [comment]
closing due to lack of interest (this PR has been open for 5 months), thanks all for comments

**Pull request closed without merging**